src/main/java/mate/academy/service/impl/AuthenticationServiceImpl.java

```diff
@@ -0,0 +1,53 @@
+package mate.academy.service.impl;
+
+import java.util.Optional;
+import mate.academy.exception.AuthenticationException;
+import mate.academy.exception.RegistrationException;
+import mate.academy.lib.Inject;
+import mate.academy.lib.Service;
+import mate.academy.model.User;
+import mate.academy.service.AuthenticationService;
+import mate.academy.service.UserService;
+import mate.academy.util.HashUtil;
+
+@Service
+public class AuthenticationServiceImpl implements AuthenticationService {
+    private static final String CANT_LOGIN_EXCEPTION_MESSAGE =
+            "Can't login. Login or password is wrong.";
+    private static final String EXISTED_EMAIL_EXCEPTION_MESSAGE =
+            "Can't register user. This email is already used";
+    private static final String EMPTY_EMAIL_EXCEPTION_MESSAGE =
+            "Email can't be empty";
+    @Inject
+    private UserService userService;
+
+    @Override
+    public User login(String email, String password) throws AuthenticationException {
+        Optional<User> userFromDbOptional = userService.findByEmail(email);
+        if (userFromDbOptional.isEmpty()) {
+            throw new AuthenticationException(CANT_LOGIN_EXCEPTION_MESSAGE);
+        }
+        User user = userFromDbOptional.get();
+        String hashedPassword = HashUtil.hashPassword(password, user.getSalt());
+        if (user.getPassword().equals(hashedPassword)) {
+            return user;
+        }
+        throw new AuthenticationException(CANT_LOGIN_EXCEPTION_MESSAGE);
```

Comment:
Create only one condition for throwing `AuthenticationException` in `login()` method. You may combine two checks: whether the user has been found by login and does passwords match.

Reply:
I tried to fix it. If I understand right we can use || in if in a way, that if we have no user, we will throw an exception, or if a user exists, we can get the user from userOptional and be sure that we won't get NPE. For now, it looks overwhelmed by text, any way, to make it better?